src/main/scala/com/virtuslab/shared_indexes/remote/S3Upload.scala

```diff
@@ -0,0 +1,128 @@
+package com.virtuslab.shared_indexes.remote
+
+import com.intellij.indexing.shared.cdn.upload.{CdnUploadAnotherEntry, CdnUploadDataEntry, CdnUploadEntry}
+import com.intellij.indexing.shared.cdn.{CdnEntry, CdnUpdatePlan, S3, S3CdnEntry}
+import com.intellij.indexing.shared.local.LocalDiskEntry
+import org.slf4j.LoggerFactory
+import software.amazon.awssdk.core.sync.RequestBody
+import software.amazon.awssdk.services.s3.model.{DeleteObjectRequest, PutObjectRequest}
+
+import java.io.{ByteArrayInputStream, InputStream}
+import java.nio.file.Files
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, TimeoutException}
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try, Using}
+
+class S3Upload(private val s3: S3, private val timeout: Duration = Duration(5, "min")) {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+  private val rootInBucket: String = s3.getRootInBucket
+  private implicit val ec: ExecutionContext = ExecutionContext.fromExecutor(s3.getDispatcher.getExecutor)
+
+  def updateS3Indexes(plan: CdnUpdatePlan): Unit = {
```

Comment:
have you considered making everything in this method async and only put Await with timeout on the last line? I.e. make it like
```
failedRemoves.map { failedRemves =>
  if (failedRemoves.nonEmpty) {
  val sb = new StringBuilder()
  sb.append(s"Failed to remove ${failedRemoves.size} entries: \n")
  sb.foreach { case (entry, e) =>
     failureSummary.append(s"  $entry: \n    ${e.getMessage}\n")
  }
  sb.toString
  } else ""
}
```
combine the results of removes and uploads in for/flatMap

it would make all actions possibly parallel

Reply:
Do you mean that removes and uploads should be executed concurrently to each other? This isn't the case in the original code, and I think it makes sense. If we executed removes concurrently with uploads, we could accidentally remove the item that we have just uploaded. Although the Jetbrains code contains a safeguard against such situations (the `CdnUpdatePlan` constructor checks whether there's an intersection between entries to remove and entries to upload), I doubt it will do the job in practice (because various subclasses of `CdnEntry` don't override `equals`).

Reply:
Ok, so let's better be safe and correct. I'd imagine that there is no delete request if one plans to replace the entry, but who knows.